### PR TITLE
Vendor `nvidia-ml-py-11.515.48`

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
Context:
- We have been using an outdated version (from 2015!) of NVIDIA's Python bindings for the [NVML library](https://developer.nvidia.com/nvidia-management-library-nvml)
- Some users report problems with the recent NVIDIA drivers on Windows resulting in BSoD (e.g. https://github.com/wandb/wandb/issues/3597 and https://github.com/wandb/wandb/issues/3819), which might be related to that fact. 
  - I tested a range of Win+NVIDIA driver version combinations but didn't manage to repro the reported issues :`(

Proposed solution:
- Vendor the Python bindings available at https://pypi.org/project/nvidia-ml-py/
  - Modifications:
    - Added `NVML_DLL_PATH` env var check to the library initialization path
    - Applied black formatting to improve readability
    - Try different versions ("_v3", "_v2", "") of the functions: `nvmlDeviceGetComputeRunningProcesses`, `nvmlDeviceGetGraphicsRunningProcesses`, and `nvmlDeviceGetMPSComputeRunningProcesses`. Reason: older driver versions don't understand _v3's that is now the only option in the lib, so I had to add this enumeration/trial loop.

- Add nightly testing on a Win+GPU executor in Circle
- Add nightly checks for new releases of vendored packages (for now, only this one) (+ anything from our requirements?), posting "interesting findings" to Slack.

Reasoning:
- https://pypi.org/project/nvidia-ml-py/ seems to be the most up-to-date version of the bindings on PyPI. 
- There is [one change](https://github.com/wandb/wandb/commit/dc79a2fe574155b09b88f63c7852a88e7df24699) that we have in our current vendored version (`NVML_DLL_PATH` env var check) that is absent in the lib. It might be not necessary, but I did find a few references to it, so it seems safer to keep it there.
- https://pypi.org/project/nvidia-ml-py seems to lack a GH repo, so I can't submit a PR.
- There is another project, https://pypi.org/project/pynvml/ that does have a GH repo that is alive, but it is less frequently updated. The main contributor seems to be from NVIDIA, so I'm not sure what its relationship with the other library is.
- I think the safest option is to vendor the "most official" bindings + add testing around it. If we find the `NVML_DLL_PATH` stuff unnecessary (check with NVIDIA!), might as well rm the vendored version and add a requirement instead.
- Thought about monkey-patching the relevant code path, but that seems risky as it's the main global function that loads the module that needs patching, it gets too funky to foresee all the possible scenarios.

Random note:
Get the latest available version of pytorch with the CUDA support that you need:
```bash
pytorch_v=`pip install --extra-index-url https://download.pytorch.org/whl/cu101 torch== 2>&1 | grep -oE '(\(.*\))' | awk -F:\  '{print$NF}' | sed -E 's/( |\))//g' | tr ',' '\n' | grep cu101 | tail -1`
```

Testing
-------
Manual nightly: ensured that polling GPU metrics on both lin/gpu and win/gpu works with the new bindings:
https://app.circleci.com/pipelines/github/wandb/wandb/14989/workflows/761624a2-c9fc-45fe-98c9-ca5b0a0c31f4

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 version: 2.1
 
 orbs:
-  win: circleci/windows@2.4.0
-  slack: circleci/slack@4.9.3
-  go: circleci/go@1.3.0
+  win: circleci/windows@5.0.0
```
Comment (dmitryduev) [inline .circleci/config.yml:4]: Probably a good idea to stay up-to-date with the orbs. This update eliminates py37 and changes the way the main user is treated in the vm.

File: tox.ini [hunk 2]
```diff
@@ -361,11 +361,11 @@ whitelist_externals =
     bash.exe
 commands =
     bash.exe -c 'mkdir -p cover-results'
-    bash.exe -c 'C:/Users/circleci/project/.tox/wincovercircle/Scripts/python -m coverage combine C:/Users/circleci/project/.tox/py37/.coverage*'
-    bash.exe -c 'C:/Users/circleci/project/.tox/wincovercircle/Scripts/python -m coverage xml'
+    bash.exe -c '~/project/.tox/wincovercircle/Scripts/python.exe -m coverage combine ~/project/.tox/py39/.coverage*'
```
Comment (dmitryduev) [inline tox.ini:364]: with the latest orb it's dynamic, something like `/c/Users/circleci.PACKER-62C74123`

File: wandb/vendor/pynvml/pynvml.py [hunk 3]
```diff
@@ -1,5 +1,16 @@
 #####
-# Copyright (c) 2011-2015, NVIDIA Corporation.  All rights reserved.
+#
+# Based on nvidia-ml-py version 11.515.48
+# Sourced from https://pypi.org/project/nvidia-ml-py/
+# Modifications:
+# - Added `NVML_DLL_PATH` env var check to the library initialization path
+# - Applied black formatting to improve readability
+# - Try different versions ("_v3", "_v2", "") of the functions:
```
Comment (dmitryduev) [inline wandb/vendor/pynvml/pynvml.py:8]: Older driver versions don't understand _v3's that is not the only option in the lib, so I had to add this enumeration/trial

**Review by kptkin — APPROVED** [review]
🥇

**Review by raubitsj — APPROVED** [review]
Looks great, thanks for going the extra mile on the testing and picking a good path vendor vs ?

**Comment by RyanYunruiYang** [comment]
Hi, this is a bit of shot in the dark and isn't exactly about the commit, but I've been trying to hunt down where the "GPU Part Number" value (one of the queryable values in nvidia-smi -q) comes from and ended up here.

I came to the conclusion that the information is in the Inforom, but couldn't figure out how to read it (checked a few versions of the c library with nm -D and couldn't find it) --> happened across the nvmlDeviceGetInforomVersion(handle, infoRomObject) function in wandb/vendor/pyvnml/pynvml.py, which suggests that someone would be able to pass in --> checked the git blame and got to this commit.

I was originally hoping that I could see nvmlDeviceGetInforomVersion was called to see how to get the infoRomObject value, but this repo doesn't call nvmlDeviceGetInforomVersion anywhere, so I'm curious if there's any of wandb's repos I should poke around for to see an example of nvmlDeviceGetInforomVersion being called?

Thanks!

**Comment by dmitryduev** [comment]
hi @RyanYunruiYang, the vendored pynvml.py is dead code, I'll delete it momentarily (thanks for the catch!). We've moved GPU monitoring to this (Rust) binary: https://github.com/wandb/wandb/tree/main/gpu_stats, see specifically https://github.com/wandb/wandb/blob/main/gpu_stats/src/gpu_nvidia.rs. We're calling into nvml using https://github.com/rust-nvml/nvml-wrapper. See if a particular query you're interested in is available here: https://github.com/rust-nvml/nvml-wrapper/blob/7e0752f331d37a62a46ef2fb6329a33fae36dd8d/nvml-wrapper/src/device.rs, we can add it on our side then.

**Pull request merged**